bundle/src/main/java/com/adobe/acs/commons/wcm/impl/RobotsServlet.java

```diff
@@ -0,0 +1,418 @@
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+@Component(service = Servlet.class, immediate = true, property = {
+        "sling.servlet.selectors=robots",
+        "sling.servlet.extensions=txt",
+        "sling.servlet.methods=GET"
+}, configurationPolicy = ConfigurationPolicy.REQUIRE)
+@Designate(ocd = RobotsServlet.RobotsServletConfig.class, factory = true)
+public final class RobotsServlet extends SlingSafeMethodsServlet {
+
+    private static final Logger log = LoggerFactory.getLogger(RobotsServlet.class);
+    private static final String ALLOW = "Allow: ";
+    private static final String USER_AGENT = "User-agent: ";
+    private static final String SITEMAP = "Sitemap: ";
+    private static final String DISALLOW = "Disallow: ";
+
+    private String externalizerDomain;
+
+    private String robotsContentsPropertyPath;
+
+    private boolean printGroupingComments;
+
+    private RobotsRuleSet rules;
+
+    @Reference
+    private Externalizer externalizer;
+
+    @Activate
+    protected void activate(RobotsServletConfig config) {
+        externalizerDomain = config.externalizer_domain();
+        robotsContentsPropertyPath = config.robots_content_property_path();
+        printGroupingComments = config.print_grouping_comments();
+
+        rules = new RobotsRuleSet(config);
+    }
+
+    @Override
+    protected void doGet(SlingHttpServletRequest request, SlingHttpServletResponse response)
+            throws ServletException, IOException {
+        response.setContentType("text/plain");
+        response.setCharacterEncoding("UTF-8");
+
+        write(request, response);
+    }
+
+    private void write(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
+        if (StringUtils.isNotBlank(robotsContentsPropertyPath)) {
+            writeFromJcrProperty(request, response);
+        } else {
+            writeFromOsgiConfig(request, response);
+        }
+    }
+
+    private void writeFromOsgiConfig(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
+        PrintWriter writer = response.getWriter();
+
+        rules.getGroups().forEach(group -> writeGroup(group, request, writer));
+
+        rules.getSitemaps().stream().map(sitemap -> buildSitemapString(sitemap, request.getResourceResolver())).forEach(writer::println);
+        if(!rules.getSitemapProperties().isEmpty()) {
+            PageManager pageManager = request.getResourceResolver().adaptTo(PageManager.class);
+            Page page = pageManager.getContainingPage(request.getResource());
```

Comment:
How do we know this will always be a page? Isnt this simply bound to some selectors/extension?

Reply:
The (or at least my) expectation, is that the config would bind it to a resource type of the page component, and it would be addressed as /path/to/page/_jcr_content.robots.txt. It is I support possible it would be bound some other resource type, but that isn't really a case I have in mind to support.